rhombus-lib/rhombus/private/amalgam/filesystem.rhm

```diff
@@ -47,5 +71,344 @@ namespace filesystem:
+        | make_one(p)
+      maybe_make_one(
+        for values(p :~ Path = root) (next: [next, ...]):
+          maybe_make_one(p)
+          p.add(next)
+      )
+    | make_one(p)
+
+  def current_force_delete_permissions:
+    #{rename-parameter}(rkt.#{current-force-delete-permissions},
+                        #'#{filesystem.current_force_delete_permissions})
+
+  fun make_temporary(
+    ~in:
+      dir :: PathString = rkt.#{find-system-path}(#'#{temp-dir}),
+    ~as:
+      copy_from :: PathString || matching(#'file) || matching(#'directory) = #'file,
+    ~make_name:
+      make_name :: Function.all_of(String -> Path.Relative,
+                                   ~name: #'#{|make_name for make_temporary|})
+        = fun (s): Path("tmp" +& s),
+    ~permissions:
+      permissions :: maybe(Int.in(0, 65535)) = #false,
+    ~replace_permissions = #false
+  ) :~ Path:
+    ~who: who
+    fun make(s = rkt.#{current-seconds}(),
+             ms = math.exact(math.truncate(rkt.#{current-inexact-milliseconds}())),
+             tries = 0):
+      let tmp: Path.add(dir, make_name(s +& ms))
+      try:
+        match copy_from
+        | #'directory:
+            rkt_error.local_error_adjust { #'#{make-directory}: who }:
+              rkt.#{make-directory}(tmp, permissions || 0o777)
+        | #'file:
+            rkt_error.local_error_adjust { #'#{open-output-file}: who }:
+              Port.Output.close(rkt.#{open-output-file}(tmp,
+                                                        ~permissions: permissions || 0o666,
+                                                        #{#:replace-permissions?}: replace_permissions))
+        | ~else:
+            rkt_error.local_error_adjust { #'#{copy-file}: who }:
+              rkt.#{copy-file}(copy_from,
+                               tmp,
+                               ~permissions: permissions,
+                               #{#:replace-permissions?}: replace_permissions)
+        tmp
+        ~catch (exn :: (Exn.Fail.Filesystem.Exists
+                          || satisfying(fun
+                                        | (exn :: Exn.Fail.Filesystem.Errno):
+                                            exn.errno.first == #'windows
+                                              && exn.errno.rest == 5
+                                              && tries < 32
+                                        | (_): #false))):
+          make(s + math.random(10), ms + math.random(10), tries + 1)
+    make()
+
+  fun make_link(to_path :: PathString,
+                path :: PathString):
+    ~who: who
+    rkt_error.local_error_adjust { #'#{make-file-or-directory-link}: who }:
+      rkt.#{make-file-or-directory-link}(to_path,
+                                         path)
+
+  fun copy(src_path :: PathString,
+           dest_path :: PathString,
+           ~recur = #false,
+           ~exists_ok = #false,
+           ~permissions: permissions :: maybe(Int.in(0, 65535)) = #false,
```

Comment:
Having a permission option that would duplicate the permissions from the source files and directories would be nice.

Reply:
Agreed.